### PR TITLE
Delete course pages changes (PTCD-427)

### PR DESCRIPTION
**Pull request by gunndabad**
Port the Delete Course journey into v2 and update the content.

image

### CODE REVIEW

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+using Dfc.CourseDirectory.Core.DataStore.CosmosDb;
+using Dfc.CourseDirectory.Core.DataStore.CosmosDb.Models;
+using Dfc.CourseDirectory.Core.DataStore.CosmosDb.Queries;
+using Dfc.CourseDirectory.Core.Models;
+using Dfc.CourseDirectory.Core.Validation;
+using Dfc.CourseDirectory.WebV2.Behaviors;
+using FluentValidation.Results;
+using MediatR;
+using OneOf;
+using DeleteCourseRunQuery = Dfc.CourseDirectory.Core.DataStore.CosmosDb.Queries.DeleteCourseRun;
```
Comment (timabell) [inline src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs:15]: Having to rename in a `using` is in my opinion a code smell, in as much as name clashes have a tendency to result in bugs and misunderstandings.

Within a single codebase I prefer slightly less concise naming to avoid having these clashes.

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+namespace Dfc.CourseDirectory.WebV2.Features.Courses.DeleteCourseRun
+{
+    public class Request : IRequest<ViewModel>
+    {
+        public Guid CourseId { get; set; }
+        public Guid CourseRunId { get; set; }
+    }
+
+    public class ViewModel : Command
```
Comment (timabell) [inline src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs:25]: I prefer the one class per file rule. I'd probably use a subfolder/namespace to keep the pieces together. Obviously one for later as we'd need to consider the whole project.

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            var (course, courseRun) = await GetCourseAndCourseRun(request.CourseId, request.CourseRunId, ukprn);
+            return await CreateViewModel(course, courseRun);
+        }
+
+        public async Task<OneOf<ModelWithErrors<ViewModel>, SuccessViewModel>> Handle(Command request, CancellationToken cancellationToken)
+        {
+            var ukprn = await GetUkprnForCourse(request.CourseId);
+            var (course, courseRun) = await GetCourseAndCourseRun(request.CourseId, request.CourseRunId, ukprn);
+
+            // User must have confirmed
```
Comment (timabell) [inline src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs:81]: Kill the comment, clear enough from the code.

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                FlexibleStartDate = courseRun.FlexibleStartDate,
+                StartDate = courseRun.StartDate,
+                VenueName = venue?.VenueName,
+                YourReference = courseRun.ProviderCourseID
+            };
+        }
+
+        private async Task<(Course Course, CourseRun CourseRun)> GetCourseAndCourseRun(Guid courseId, Guid courseRunId, int ukprn)
+        {
+            (await _cosmosDbQueryDispatcher.ExecuteQuery(new GetCoursesByIds()
```
Comment (timabell) [inline src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs:130]:
```suggestion
            (await _cosmosDbQueryDispatcher.ExecuteQuery(new GetCoursesByIds
```
remove redundant parenthethethis on object instantiation

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            if (!providerId.HasValue)
+            {
+                throw new ResourceDoesNotExistException(ResourceType.Course, courseId);
+            }
+
+            var providerInfo = await _providerInfoCache.GetProviderInfo(providerId.Value);
+            return providerInfo.Ukprn;
+        }
+
+        Guid IRequireUserCanAccessCourse<Request>.GetCourseId(Request request) => request.CourseId;
```
Comment (timabell) [inline src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs:164]: For the sake of my understanding, what does this do? I'm guessing access control, but I can't guess how it fits into the flow etc.

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cshtml [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        <govuk-error-summary-item asp-for="Confirm" />
+    </govuk-error-summary>
+}
+
+<div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+        <form asp-action="DeleteCourseRun" asp-route-courseId="@Model.CourseId" asp-route-courseRunId="@Model.CourseRunId" asp-route-returnUrl="@returnUrl">
+            <h1 class="govuk-heading-xl">Delete course</h1>
+
+            <p class="govuk-body">Deleting this course will remove it from Publish to the course directory and National Careers Service, Find a course.</p>
```
Comment (timabell) [inline src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cshtml:22]: Long lines make diffs hard to review in github. Sad but true. Suggest wrapping this file's long lines.

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cshtml [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                            Are you sure?
+                        </h2>
+                    </govuk-checkboxes-fieldset-legend>
+                </govuk-checkboxes-fieldset>
+                <govuk-checkboxes-item value="@(true)">
+                    Yes - I want to delete the course
+                </govuk-checkboxes-item>
+            </govuk-checkboxes>
+
+            <govuk-button type="submit" class="govuk-button--warning" id="pttcd-courses__delete-course__btn">Delete</govuk-button>
```
Comment (timabell) [inline src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cshtml:66]: I wonder if this needs visually hidden context for what is being deleted. Probably would benefit from some a11y testing.

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cshtml [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    <script type="text/javascript">
+        function enableDeleteBtn() {
+            $('#pttcd-courses__delete-course__btn').attr(
+                'disabled',
+                $('#Confirm').is(':checked') ? null : 'disabled');
+        }
+
+        // timeout is required otherwise the _checked_ property check doesn't return the correct value
+        // (maybe GDS scripts need chance to run?)
+        setTimeout(enableDeleteBtn, 100);
```
Comment (timabell) [inline src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cshtml:83]: Makes me cringe but don't have any bright ideas right now.

File: tests/Dfc.CourseDirectory.Testing/DataStore/CosmosDb/Queries/CreateCourse.cs [hunk 9]
```diff
@@ -47,5 +47,6 @@ public class CreateCourseCourseRun
         public CourseAttendancePattern AttendancePattern { get; set; }
         public bool? National { get; set; }
         public IEnumerable<string> Regions { get; set; }
+        public string ProviderCourseId { get; set; }
```
Comment (timabell) [inline tests/Dfc.CourseDirectory.Testing/DataStore/CosmosDb/Queries/CreateCourse.cs:50]: It's not immediately clear to me why this change was needed from the code or the commit message. Suggest amending the relevant commit message to add reason for needing to extend this.
I could see it was something test related, but not why the tests now needed this when they didn't before.

**Review by timabell — COMMENTED** [review]
some comments, gotta run, will pick up the rest of the diff later

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs (hunk 2 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs:25]: Yeh I think we need to have a separate PR to update everything if we're going to change it. Gives me chance to convince you that this way is better ;-)

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cshtml (hunk 7 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cshtml:66]: This lines up with what's in the prototype

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs (hunk 1 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs:15]: If I change the naming of any of these types then they will be inconsistent with everything else. I'm not wild about the clash either but I'm more concerned about being inconsistent. Perhaps one to look at holistically across the project?

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs (hunk 5 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs:164]: If a handler implements `IRequireUserCanAccessCourse<TRequest>` then it will get the course authorization rules applied to it (defined by https://github.com/SkillsFundingAgency/dfc-coursedirectory/blob/main/src/Dfc.CourseDirectory.WebV2/Behaviors/RequireUserCanAccessCourseBehavior.cs)

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/Courses/DeleteCourseRunTests.cs [hunk 10]
```diff
@@ -0,0 +1,534 @@
...
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+
+            var doc = await response.GetDocument();
+            Assert.Equal("Test Venue", doc.GetSummaryListValueWithKey("Course location"));
+        }
+
+        [Fact]
+        public async Task Get_ValidRequestCourseRunWithNoVenue_DoesNotRenderLocationRow()
```
Comment (timabell) [inline tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/Courses/DeleteCourseRunTests.cs:156]: It doesn't jump out at me what the key change in data is compared to the above test that results in rendering vs not rendering. Maybe a code comment on the relevant item(s) would suffice.

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/Courses/DeleteCourseRunTests.cs [hunk 11]
```diff
@@ -0,0 +1,534 @@
...
+            await User.AsProviderUser(providerId, ProviderType.FE);
+
+            // Act
+            var response = await HttpClient.SendAsync(request);
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+
+            var doc = await response.GetDocument();
+            Assert.Equal("01/04/2020", doc.GetSummaryListValueWithKey("Start date"));
```
Comment (timabell) [inline tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/Courses/DeleteCourseRunTests.cs:336]: Is the view that generates this locale-proofed?

File: src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+        }
+
+        public async Task<ViewModel> Handle(Request request, CancellationToken cancellationToken)
+        {
+            var ukprn = await GetUkprnForCourse(request.CourseId);
+            var (course, courseRun) = await GetCourseAndCourseRun(request.CourseId, request.CourseRunId, ukprn);
+            return await CreateViewModel(course, courseRun);
+        }
+
+        public async Task<OneOf<ModelWithErrors<ViewModel>, SuccessViewModel>> Handle(Command request, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs:76]: Name `request` comes from the interface declaration

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/Courses/DeleteCourseRunTests.cs (hunk 10 above)
Comment (gunndabad) [reply tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/Courses/DeleteCourseRunTests.cs:156]: Added explicit `venueId: null` to the `CreateCourse` method
